[gen-l10n] Fixes named and positional parameter issue with NumberFormat when type is specified

ArishSultan opened the pull request:
Fixes incorrect code generation in the following cases.
1)- When a format having optional positional parameters i.e. decimalPattern, scientificPattern, or percentPattern is selected.
2)- When an optionalParameter with String value was specified in .arb file, for example [name] in [NumberFormat.currency]

List of Issue
1)- 67217


## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

ArishSultan committed:
:heavy_plus_sign: Added a new template for [NumberFo
rmat] with positional parameter(s)

ArishSultan committed:
:heavy_plus_sign: Added a new template for [NumberFo
rmat] with positional parameter(s)

ArishSultan committed:
:hammer: Renamed [numberFormatTemplate] to [numberFormatNamedTemplate]

ArishSultan committed:
:hammer: Renamed [numberFormatTemplate] to [numberFormatNamedTemplate]

ArishSultan committed:
[gen-l10n] Fixed generation error for namedParameters and optionalParameters

ArishSultan committed:
[gen-l10n] Fixed generation error for namedParameters and optionalParameters

ArishSultan committed:
[gen-l10n] Fixed string value generation for optional parameter

ArishSultan committed:
[gen-l10n] Fixed string value generation for optional parameter

ArishSultan committed:
[gen-l10n-test] Added new tests for `decimalPattern`, `percentPattern`, `scientificPattern` and formats with optional parameters.

ArishSultan committed:
[gen-l10n-test] Added new tests for `decimalPattern`, `percentPattern`, `scientificPattern` and formats with optional parameters.

drekka commented:
Looking forwards to this :-)

shihaohong commented on packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart:
```diff
@@ -410,6 +414,54 @@ void main() {
       }
     }
   },
+  
```
A bunch of trailing whitespace analysis failures: 
```
▌10:29:03▐ STARTING ANALYSIS
▌10:29:03▐ runtimeType in toString...
▌10:29:03▐ Unexpected binaries...
▌10:29:03▐ RUNNING: cd .; git ls-files -z
▌10:29:03▐ ELAPSED TIME: 0.032s for git ls-files -z in .
▌10:29:03▐ Trailing spaces...
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
/b/s/w/ir/k/flutter sdk/packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart:417: trailing U+0020 space character
/b/s/w/ir/k/flutter sdk/packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart:432: trailing U+0020 space character
/b/s/w/ir/k/flutter sdk/packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart:443: trailing U+0020 space character
/b/s/w/ir/k/flutter sdk/packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart:454: trailing U+0020 space character
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
```

ArishSultan replied:
how did you run this analysis?

shihaohong replied:
Run `flutter analyze --flutter-repo`. If that doesn't work, run `flutter analyze` in `flutter/packages/flutter_tools`.

shihaohong commented on packages/flutter_tools/lib/src/localizations/gen_l10n_templates.dart:
```diff
@@ -112,12 +112,17 @@ abstract class @(class) {
-const String numberFormatTemplate = '''
-    final intl.NumberFormat @(placeholder)NumberFormat = intl.NumberFormat.@(format)(
+const String numberFormatPositionalTemplate = '''
+    final @(placeholder)NumberFormat = intl.NumberFormat.@(format)(localeName);
+    final @(placeholder)String = @(placeholder)NumberFormat.format(@(placeholder));
```
Follow Flutter style guide and specify the type of each variable. It seems that we unfortunately do not catch this in generated code. I'll file an issue.
```suggestion
    final intl.NumberFormat @(placeholder)NumberFormat = intl.NumberFormat.@(format)(localeName);
    final String @(placeholder)String = @(placeholder)NumberFormat.format(@(placeholder));
```

ArishSultan replied:
Google's style recommendation has a section named `Avoid using var and dynamic`.
I think I got confused because the effective dart style says do not specify type where it can be inferred from RHS value.

I thought it was redundant to specify types because it can be inferred by RHS at the time of initialization.

shihaohong replied:
In the framework, we have the `always_specify_types` lint enabled: https://github.com/flutter/flutter/blob/master/analysis_options.yaml#L56

ArishSultan replied:
Ok, will keep it in mind.

ArishSultan replied:
Can't we just shorten it? I mean it can be reduced to a single line.

```
final intl.NumberFormat @(placeholder)NumberFormat = intl.NumberFormat.@(format)(localeName);
final String @(placeholder)String = @(placeholder)NumberFormat.format(@(placeholder));
```

can be reduced to

```
final String @(placeholder)String = intl.NumberFormat.@(format)(localeName).format(@(placeholder));
```

shihaohong replied:
I originally did it this way because `intl.NumberFormat.@(format)(localeName).format(@(placeholder));` is incredibly confusing to read. The generated code might be simpler to read, but it definitely looks weird given that there's interpolated strings, along with actual function arguments

shihaohong commented on packages/flutter_tools/lib/src/localizations/gen_l10n_templates.dart:
```diff
@@ -112,12 +112,17 @@ abstract class @(class) {
+''';
+
+const String numberFormatNamedTemplate = '''
+    final @(placeholder)NumberFormat = intl.NumberFormat.@(format)(
```
```suggestion
    final intl.NumberFormat @(placeholder)NumberFormat = intl.NumberFormat.@(format)(
```

shihaohong commented on packages/flutter_tools/lib/src/localizations/gen_l10n_templates.dart:
```diff
@@ -112,12 +112,17 @@ abstract class @(class) {
       @(parameters)
     );
-    final String @(placeholder)String = @(placeholder)NumberFormat.format(@(placeholder));
+    final @(placeholder)String = @(placeholder)NumberFormat.format(@(placeholder));
```
```suggestion
    final String @(placeholder)String = @(placeholder)NumberFormat.format(@(placeholder));
```

shihaohong commented on packages/flutter_tools/lib/src/localizations/gen_l10n.dart:
```diff
@@ -146,13 +146,26 @@ String generateNumberFormattingLogic(Message message) {
-        .replaceAll('@(format)', placeholder.format)
-        .replaceAll('@(parameters)', parameters.join(',    \n'));
+
+      if (placeholder.hasNumberFormatWithParameters) {
```
nice!

shihaohong commented on packages/flutter_tools/lib/src/localizations/gen_l10n.dart:
```diff
@@ -146,13 +146,26 @@ String generateNumberFormattingLogic(Message message) {
+          if (parameter.value is num) {
+            return '${parameter.name}: ${parameter.value}';
+          } else {
+            return '${parameter.name}: ${_generateStringParameterValue(
```
Could you explain why `_generateStringParameterValue` is necessary here? It could be helpful to add a comment above the method explaining this. 

It also seems to address some case unrelated to the number formatting? If this is indeed unrelated (but important), make sure to add a test for this to make sure that this doesn't regress

ArishSultan replied:
Some formats have String parameters i.e. NumberFormat.currency has a parameter `name`. If some user passes a value with single-quote (') or double-quotes (") in the value. such as.

{
  "name": "asd'"
}

then this function will generate an optimized value i.e.
1)- `asd'` will be converted to "asd'"
2)- `asd"` will be converted to 'asd"'
3)- `asd"'` this one contains both single quote and double quotes so it will be converted as 'asd"\''

shihaohong replied:
A similar function already exists: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/localizations/localizations_utils.dart#L267

Could you use this one instead and add a few tests covering these examples? Thanks!

ArishSultan replied:
Ok, I didn't know about it, will use it now.

shihaohong replied:
Don't forget to add tests for make sure that the special cases in the string are accounted for.

engine-flutter-autoroll committed:
Roll Plugins from 815beaff69c6 to 35847e4734d4 (2 revisions) (#75038)

* 97646369b Remove amirh from CODEOWNERS (flutter/plugins#3484)

* 35847e473 [local_auth] Fix incorrect switch fallthrough (flutter/plugins#3473)

engine-flutter-autoroll committed:
04b0b4b13 [path_provider_windows] Resolve FFI stabilization changes (flutter/plugins#3485) (#75136)

goderbauer committed:
Clarify performance overlay in performance issue template (#75024)

goderbauer committed:
Bump flutter_gallery_assets (#75026)

engine-flutter-autoroll committed:
Roll Engine from ce14c8a84142 to 833d9953a742 (6 revisions) (#75046)

ArishSultan committed:
:hammer: Removed trailing whitespaces

ArishSultan committed:
:hammer: Removed trailing whitespaces

ArishSultan committed:
:hammer: replaced [_generateStringParameterValue] with [generateString]

ArishSultan committed:
:hammer: replaced [_generateStringParameterValue] with [generateString]

ArishSultan committed:
:hammer: Specified variable types in NumberFormat templates.

ArishSultan committed:
:hammer: Specified variable types in NumberFormat templates.

Anas35 committed:
removed new keyword (#75049)

goderbauer committed:
Migrate driver tests in example/ to NNBD (#75022)

dnfield committed:
Reland integration_test NNBD (#75074)

* Revert "Revert "NNBD integration_test (#74922)" (#75030)"

This reverts commit 87b043658bc3eb16cf1032cbc9b0ba6484271cea.

* fix bad merge issues

goderbauer committed:
Migrate off e2e (#75029)

jmagman committed:
Revert "Migrate driver tests in example/ to NNBD (#75022)" (#75161)

This reverts commit 50a8ecb7d6cd1fe34bd7982f463909fb649e98e8.

chunhtai committed:
fixes preceision error during SliverGridRegularTileLayout (#74851)

jmagman committed:
Split tools_tests subshards into separate shards to support sub-sub-sharding (#75033)

creativecreatorormaybenot committed:
Fix formatting in rendering/box (#75102)

jmagman committed:
Avoid iOS local networking dialog in integration tests (#75163)

Commit:
Remove scrollability by default (#75138)

engine-flutter-autoroll committed:
Roll Engine from 833d9953a742 to 9a2795ae17ca (20 revisions) (#75140)

Hixie committed:
Unpin mysteriously pinned packages (#74885)

robert-ancell committed:
Make Linux apps non-unique by default. (#74959)

This stops them attempting to own a D-Bus name, which may not be possible if the
app is sandboxed (e.g. in a Snap). It's also probably what most developers would
expect is the correct behaviour by default.

LongCatIsLooong committed:
Move caret/highlight painting to custom painters (#72828)

xu-baolin committed:
Reland "Update PopupMenuButton to match Material Design spec" (#74620)

jmagman committed:
Dump logs on failing devicelab test to recipe artifact location (#74378)

jmagman committed:
Replace context FakeUsage with TestUsage (#75170)

Hixie committed:
Remove the timeout when launching DevTools (#74859)

engine-flutter-autoroll committed:
Roll Engine from 9a2795ae17ca to aff8465afd80 (5 revisions) (#75174)

shihaohong commented:
Still waiting on tests for the use of `generateString` to make sure that doesn't regress. Also, the `Linux flutter_plugins` check is failing because of a CI issue, so you'll need to rebase or merge with the `master` branch to get that passing again.

jlubeck committed:
Adds First/Last buttons to PaginatedDataTable (#73964)

HansMuller committed:
Added ButtonStyle.splashFactory and NoSplash.splashFactory (#74949)

jmagman committed:
Adopt subshard naming convention in build_tests (#75179)

jmagman committed:
Replace MockConfig with TestConfig (#75044)

engine-flutter-autoroll committed:
Roll Plugins from 04b0b4b13f59 to 31a8b5c3d5be (2 revisions) (#75177)

engine-flutter-autoroll committed:
Roll Engine from aff8465afd80 to e59457b81854 (4 revisions) (#75181)

engine-flutter-autoroll committed:
30721bed3 Automatically add platform labels (flutter/plugins#3487) (#75184)

engine-flutter-autoroll committed:
bb2bbde4d Roll Skia from fde740c7f26e to a6c692b8844d (7 revisions) (flutter/engine#24118) (#75186)

engine-flutter-autoroll committed:
a87497f13 Add plugin issue query to README (flutter/plugins#3493) (#75190)

YeungKC committed:
Improve Dropdown Menu (#74906)

ArishSultan committed:
[gen-l10n] :white_check_mark: added test for parameter having special characters in string value.

ArishSultan committed:
Merge remote-tracking branch 'origin/issue-67217' into issue-67217

# Conflicts:
#	packages/flutter_tools/test/integration.shard/gen_l10n_test.dart
#	packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart

shihaohong commented:
Looks like the merge/rebase with master went unexpectedly like with the previous PR. I can take over and submit the fix if you don't mind. My main concern is I don't want to keep other devs waiting on the fix. Can you tell me what steps you follow that cause this to happen? I could also try to debug the source of your problem as well

shihaohong commented:
I created a PR at https://github.com/flutter/flutter/pull/75209, which contains your commit so your commit history is preserved as a contributor.

ArishSultan commented:
I updated the repo using fetch upstream and then rebased my branch on master. what did I do wrong?

shihaohong commented:
The way that I do it is from the branch I'm working on, I use `git pull upstream master`, after setting my remote `upstream` to be `git@github.com:flutter/flutter.git`. Then, I'd fix any merge conflicts and then push that to GitHub. What exact commands are you using?

ArishSultan commented:
I used git fetch upstream. I think I should have used pull instead of fetch
